core/src/org/samtools/htsjdk/core/io/ByteSource.java

```diff
@@ -0,0 +1,107 @@
+package org.samtools.htsjdk.core.io;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Interface for classes that produce binary data (i.e. bytes) that can be decoded
+ * into other datatypes.
+ *
+ * Methods should throw exceptions if:
+ * - The underlying data source/stream/channel is closed
+ * - There are insufficient bytes available (after blocking) to read a single item of the datatype requested
+ *
+ * NOTES:
+ *   - Almost all the methods here could have default implementations that read via one of the readBytes() methods.
+ *   - Should think through how to deal with unsigned types
+ *   - Need to think how this would work with BlockCompressed data
+ */
+public interface ByteSource extends AutoCloseable {
+    /** Returns a description of the underlying source of the bytes, e.g. a path, stdin, a URL etc. */
+    String getDescription();
+
+    // Methods related to buffering/repositioning/skipping etc.
+    /** Returns true if the source has a known finite length, false otherwise. */
+    boolean hasLength();
+
+    /** Returns the length if {@link #hasLength()} is true, otherwise throws an exception. */
+    default long length() { throw new UnsupportedOperationException(); }
+
+    /** Returns the offset of the next byte that will be read by any operation. */
+    long position();
+
+    /** Returns true if the source can skip bytes without reading them. */
+    boolean canNavigateForward();
+
+    /** Returns true if the source can always be repositioned behind the current position. Note that a buffered
+     * source may allow limited backwards navigation within the buffered data, while returning false here. */
+    boolean canNavigateBackwards();
```

Comment:
Instead of returning false here for limited backwards navigation, we can return number of bytes that can be "safely" navigate backwards:

* Buffered data: buffer-size (it could be big for pipes to discover formats, for example)
* Random access data: return the current position (in principle, it can be completely reset)
* Unbuffered data: returns 0.

Reply:
And related with this method, we can also include a `canReset` and `reset` to come back to the first position in the file (as a sugar syntax for `position(0)` - this will be just nice to have (I can imagine several parts of the code calling the method to return the number of bytes that can be navigate backwards, compare to the lenght and if they are equal position the stream in 0).

Reply:
I did think about this @magicDGS I think the problem is that this doesn't work great for cases where you're at `position(0)` already, and want to ask "if I read N bytes to do format detection, can I go back to 0, or do I need to buffer those bytes myself".

Reply:
True. Maybe for random access data it should return the maximum long value and for buffered data the buffer size. This would behave correctly in your case of format detection: even in position 0, the answer is true with that API definition.